include/up-cpp/utils/Expected.h

```diff
@@ -22,77 +22,114 @@ namespace uprotocol::utils {
+// nor earlier versions.8 Could be the issue is specific to non-const versus
+// const. The loss of this part of the implementation means swapping is not
+// being done, and useless copies are being done. This should be reexamined if
+// there are use cases where the expected value ends up being a large object.
+//
+//////////////////////////////////////////////////////////////////////////////////////////////
+
 /// @name Temporary substitute for std::expected
 /// @remarks See the reference for std::expected:
 ///          https://en.cppreference.com/w/cpp/utility/expected
 ///          No further documentation is provided in this file.
 /// @{
-template <typename E>
-struct BadExpectedAccess;
+struct BadExpectedAccess : public std::runtime_error {
+	BadExpectedAccess() : std::runtime_error("generic BadExpectedAccess") {}
+	BadExpectedAccess(const char* arg) : std::runtime_error(arg) {}
+};
 
-template <>
-struct BadExpectedAccess<void> : public std::exception {};
+/// @brief Required tagging type for cases where expected and unexpected type
+/// are identical.
+template <typename T>
+class Unexpected {
+	T t;
 
-template <typename E>
-struct BadExpectedAccess : public BadExpectedAccess<void> {};
+public:
+	Unexpected(const T& arg) { t = arg; }
+	const T& value() const { return t; }
+};
 
 /// @brief A stripped-down version of std::expected from C++23.
 template <typename T, typename E>
 struct Expected {
-	using value_type = T;
-	using error_type = E;
-
-	constexpr Expected();
-	constexpr Expected(const Expected& other);
-	constexpr Expected(Expected&& other) noexcept;
-
-	constexpr explicit Expected(T&& v);
-	constexpr explicit Expected(E&& e);
-
-	Expected& operator=(const Expected& other);
-	Expected& operator=(Expected&& other);
-
-	Expected& operator=(const T& other);
-	Expected& operator=(T&& other);
-
-	Expected& operator=(const E& other);
-	Expected& operator=(E&& other);
-
-	constexpr explicit operator bool() const noexcept;
-	constexpr bool has_value() const noexcept;
-
-	constexpr T& value() &;
-	constexpr const T& value() const&;
-	constexpr T&& value() &&;
-	constexpr const T&& value() const&&;
-
-	constexpr const E& error() const& noexcept;
-	constexpr E& error() & noexcept;
-	constexpr const E&& error() const&& noexcept;
-	constexpr E&& error() && noexcept;
-
-	constexpr T value_or(T&& default_value) const&;
-	constexpr T value_or(T&& default_value) &&;
-
-	template <typename F>
-	constexpr auto and_then(F&& f) &;
-	template <typename F>
-	constexpr auto and_then(F&& f) const&;
-	template <typename F>
-	constexpr auto and_then(F&& f) &&;
-	template <typename F>
-	constexpr auto and_then(F&& f) const&&;
-
-	template <typename F>
-	constexpr auto or_else(F&& f) &;
-	template <typename F>
-	constexpr auto or_else(F&& f) const&;
-	template <typename F>
-	constexpr auto or_else(F&& f) &&;
-	template <typename F>
-	constexpr auto or_else(F&& f) const&&;
+public:
+	constexpr Expected() { new (&yay) T(); }
```

Comment:
If T has any non-trivial destructor, this would be extremely dangerous as the destructor for the original `yay` never runs (its constructor _does_ run). Why aren't values being assigned or moved in the standard way? What is the intent behind using _placement new_ in this way? Isn't it typically intended only for construction into a pre-allocated buffer and not for reassigning values?

I cannot see a way that this _wouldn't_ result in memory leaks.